### PR TITLE
[sparsity] Base sparsity level scheduler class

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* **#59770 [sparsity] Base sparsity level scheduler class**
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* #58704 [sparsity] Sparsifier class
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements the base scheduler class for changing the sparsity levels in the sparsifier.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D29070603](https://our.internmc.facebook.com/intern/diff/D29070603)

### CODE REVIEW

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

File: torch/ao/sparsity/scheduler/base_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    def __repr__(self):
+        format_string = self.__class__.__name__ + ' ('
+        format_string += '\n'
+        format_string += 'Sparsifier {0}\n'.format(self.sparsifier)
+        format_string += '    {0}: {1}\n'.format('base_sl', self.base_sl)
+        format_string += ')'
+        return format_string
+
+    def step(self, epoch=None):
+        # Raise a warning if old pattern is detected
```
Comment (raghuramank100) [inline torch/ao/sparsity/scheduler/base_scheduler.py:103]: Since we do not have the old pattern, can we simplify the logic here?

File: torch/ao/sparsity/scheduler/base_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        # https://github.com/pytorch/pytorch/issues/20124
+        self._step_count += 1
+
+        class _enable_get_sl_call:
```
Comment (raghuramank100) [inline torch/ao/sparsity/scheduler/base_scheduler.py:107]: Not sure if this context manager is needed.

File: torch/ao/sparsity/scheduler/base_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            raise TypeError('{} is not an instance of torch.ao.sparsity.BaseSparsifier'.format(
+                type(sparsifier).__name__))
+        self.sparsifier = sparsifier
+
+        # Initialize epoch and base sparsity levels
+
+        self.base_sl = [group['sparsity_level'] for group in sparsifier.module_groups]
+        self.last_epoch = last_epoch
+
+        # Following https://github.com/pytorch/pytorch/issues/20124
```
Comment (raghuramank100) [inline torch/ao/sparsity/scheduler/base_scheduler.py:22]: The logic here is hard to follow. Is this the only way to implement this?

File: torch/ao/sparsity/scheduler/base_scheduler.py (hunk 2 above)
Comment (z-a-f) [reply torch/ao/sparsity/scheduler/base_scheduler.py:107]: Basically, this context manager is a guardrail to prevent the users from calling the `get_sl` manually. The reason I wouldn't want them to is because there is a separate function `get_last_sl`, which shows the latest computed sl. Otherwise, the sl will be reported incorrectly, depending on which epoch you are calling it on.

File: torch/ao/sparsity/scheduler/base_scheduler.py (hunk 3 above)
Comment (z-a-f) [reply torch/ao/sparsity/scheduler/base_scheduler.py:22]: I put as many comments in here. But in general, I agree this is a little convoluted. I will change the first comment to make sure it is more clear.

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

File: test/ao/sparsity/test_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+class ImplementedScheduler(BaseScheduler):
+    def get_sl(self):
+        if self.last_epoch > 0:
+            return [group['sparsity_level'] * 0.5
+                    for group in self.sparsifier.module_groups]
+        else:
+            return list(self.base_sl)
+
+
+class TestBaseScheduler(TestCase):
```
Comment (raghuramank100) [inline test/ao/sparsity/test_scheduler.py:17]: Please add tests for verifying (1) Sparsifier is called before scheduler. (2) get_sl method behaves as expected.

**Review by raghuramank100 — APPROVED** [review]

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29070603).
